### PR TITLE
Add ActuatorCurrentLimit message.

### PR DESCRIPTION
**Pull request by adolfo-rt**
Message type for setting the current limit for a group of actuators.

This is part of a larger set of interfaces that aim at exposing current limiting capabilities, and is yet another stab at what PRs #9 and #10 have proposed. Consider this a further iteration of these efforts.
#### Rationale
- A separate message type is convenient in the sense that it can be readily streamed via topic pub-sub, and can also be included in service req-rep, if needed.
- The message fields being arrays allows to specify current limits to one or more actuators, and mimics the structure of other ROS messages, like `sensor_msgs::JointState` and `trajectory_msgs::JointTrajectory`.
#### The big picture

The idea is to have the following interfaces (this PR deals with the first one only):
- **ROS message** `pal_control_msgs::ActuatorCurrentLimit`
- **ros_control types**
  - **Handle** `pal_ros_control::CurrentLimitHandle`
  
  ``` c++
  class CurrentLimitHandle
  {
  public:
    CurrentLimitHandle();
    CurrentLimitHandle(const std::string& name, double* curr_lim);
  
    std::string getName() const;
    double      getCurrentLimit() const;
    void        setCurrentLimit(double curr_lim);
  };
  ```
  - **Interface** `pal_ros_control::CurrentLimitInterface`, which does not claim its resources
- **ros_control controller**
  - `pal_ros_controllers::CurrentLimitController`
  - Is configured from a yaml file with this structure
  
  ``` yaml
    head_current_limit_controller:
    type: "pal_ros_controllers/CurrentLimitController"
    actuators:
      - head_1_motor
      - head_2_motor
  ```
  - Takes `pal_control_msgs::ActuatorCurrentLimit` commands  from a topic interface
#### Feedback needed

Apart from giving feedback on this particular PR, please also comment on the overall naming and structure of the other interfaces, as they should be coherent.

One particular aspect that I'm not decided on, is that the message name makes it explicit that we're dealing with actuators, while the `ros_control` side drops this identifier for brevity. What are your preferences?.

If we reach some consensus here, I can push to our private repo an implementation of the `ros_control` parts to our internal repo, while remaining compatible with existing code until we determine a migration path, if any.

@awesomebytes, @bmagyar, @lucamarchionni, @po1, @v-lopez

### CODE REVIEW

**Comment by po1** [comment]
What is the difference in working with _actuators_ instead of _joints_?

The looks fine and functional, maybe a bit of inline documentation would help.
Also, you named it `ActuatorCurrentLimit`, singular. It looks like it is supposed to contain many of these limits. Wouldn't the plural be more appropriate?

**Comment by adolfo-rt** [comment]
> What is the difference in working with actuators instead of joints?

When you have a transmission that is not a simple reducer, like a differential, it just doesn't make sense to map effort limits (or current limits) between joint and actuator spaces, while it does make sense to map efforts (or currents). Consider a transmission given by this expression, where `f` denotes a joint effort and `t` an actuator effort:

```
f1 = t1 + t2
f2 = t1 - t2
```

##### Example 1

If we want to have joint effort limits of `f1 = 0.5` and `f2 = 0.1`, we would have to limit actuators to `t1 = 0.3` and `t2 = 0.2`, so far so good.

##### Example 2

If we want to have joint effort limits of `f1 = 0.1` and `f2 = 0.5` (the opposite of example 1), we would have to limit actuators to `t1 = 0.3` and `t2 = -0.2`. Considering that effort/current limits are non-negative values, `t2 = -0.2` is an impossible setting.

Now, if we're controlling efforts, and not effort limits, it's perfectly fine to exert a negative effort in one actuator.

> The looks fine and functional, maybe a bit of inline documentation would help.

OK

> Also, you named it ActuatorCurrentLimit, singular. It looks like it is supposed to contain many of these limits. Wouldn't the plural be more appropriate?

True, I was just following suit of the `sensor_msgs/JointState` message. In fact, this also applies to the member `current_limit`, which could also be plural.

In ROS, there is no clear consensus about usage of singular/plural in messages. For example, `sensor_msgs/JointState` has a member `float64[] position` (array, singular), while  `trajectory_msgs/JointTrajectoryPoint` has a member `float64[] positions` (array, plural). Maybe older ROS messages follow the wrong convention, but have remained unchanged for compatibility reasons.

**Comment by po1** [comment]
I now understand the difference, thanks.
So this means that if you have a differential transmission, there will be (as you pointed out) limiting configurations that will not be possible. While I understand the theoretical limit, I find this practically unsatisfactory. For example, there might be a legitimate reason to want full stiffness on `f1` but none at all on `f2` (or the other way around), and the proposed architecture does not seem to address that. Moreover, this means that the user of this message will have to know the details of the actual mechanical implementation of joints.

I don't mean to change anything or to delay merging this. This is evidently useful as it is.

> In ROS, there is no clear consensus about usage of singular/plural in messages.

That's true, as it seems. Then I guess anything is fine, as long as we stick to something. It is up to you to decide whether you want to follow `sensor_msgs` or `trajectory_msgs`.

**Comment by adolfo-rt** [comment]
> While I understand the theoretical limit, I find this practically unsatisfactory. For example, there might be a legitimate reason to want full stiffness on f1 but none at all on f2 (or the other way around), and the proposed architecture does not seem to address that.

Actuator current limits are a pretty low-level thing. If one wants to control efforts or stiffness, then an alternative method should be used. For instance, if we do the position PID loop in software, and at the joint-level; we would produce joint torques that can be manipulated more effectively. The compromise would be that we would need to send control commands more often, and require  shorter control delays, and better velocity estimates.

> Moreover, this means that the user of this message will have to know the details of the actual mechanical implementation of joints.

Yes, having to explicitly know about actuators in an otherwise joint-level controller is undesired.

**Comment by adolfo-rt** [comment]
> That's true, as it seems. Then I guess anything is fine, as long as we stick to something. It is up to you to decide whether you want to follow sensor_msgs or trajectory_msgs.

I made the message fields plural, but not the message name, as `ActuatorCurrentLimits` sounds like an actuator can have multiple limits.

Inline message doc is there as well.

File: pal_control_msgs/CMakeLists.txt [hunk 1]
```diff
@@ -1,13 +1,17 @@
...
 project(pal_control_msgs)
 
-find_package(catkin REQUIRED COMPONENTS message_generation actionlib_msgs)
+find_package(catkin REQUIRED COMPONENTS message_generation
+                                        std_msgs
+                                        actionlib_msgs)
 
 include_directories(include)
 
+add_message_files(DIRECTORY msg FILES ActuatorCurrentLimit.msg)
```
Comment (po1) [inline pal_control_msgs/CMakeLists.txt:10]: Please update this line to reflect the new message name

Comment (adolfo-rt) [inline pal_control_msgs/CMakeLists.txt:10]: Oops

**Comment by po1** [comment]
@adolfo-rt it looks like this PR needs to be rebased

@awesomebytes, @bmagyar, @lucamarchionni, @v-lopez other thoughts?

**Comment by lucamarchionni** [comment]
I like the proposal made by Adolfo for the message and the interfaces for
ros_control and I agree in merging the PR.
Managing actuators current limits is a trick to emulate variable stiffness
and for this reason the user should now how joints are mapped to actuators.
When we will have force/torque control, things will change.

On Friday, August 1, 2014, Paul Mathieu notifications@github.com wrote:

> @adolfo-rt it looks like this PR needs to be rebased
> 
> @awesomebytes, @bmagyar, @lucamarchionni, @v-lopez other thoughts?
> 
> —
> Reply to this email directly or view it on GitHub.<
> https://ci6.googleusercontent.com/proxy/Mexg55OhgA6xEFFNAYbgRim4C4YfHQvJLo02YKGX1XJEAMLKpX6rDUNI9XNY8v3i6nYwjqqd9gNoM9KMCkeU2MnKqYHLbnNUtF53MFUDyJMgQ7fc9WJ5d-vm97sJKxlGlhG2MwQgNBim9hAevCpU11d2T4x3mqZqN4e_BFQDU469-RkQRKSK7WkVRmRZZi8gHVG1yCVhJcqaNBPHl2-jeWhcAd2ZL0klUp5YGWbJUJWbD2dnYrIlsXxj_k9MD38pKrCf9KgYw1VCQHxUKn1H_Wj9jeyY8g=s0-d-e1-ft#https://github.com/notifications/beacon/1038209__eyJzY29wZSI6Ik5ld3NpZXM6QmVhY29uIiwiZXhwaXJlcyI6MTcyMjUyNjc1MiwiZGF0YSI6eyJpZCI6MzU4NzM2Mzh9fQ==--feac45f428ce98229d6ab605a3c1906c9f10ca2b.gif

## 

_Luca Marchionni_
_Robotics Engineer_
_luca.marchionni@pal-robotics.com luca.marchionni@pal-robotics.com_
www.pal-robotics.com

Pal Robotics, S.L.
c/ Pujades 77-79, 4º4ª
08005 Barcelona
Spain
Tel      +34 93 414 53 47
Fax     +34 93 209 11 09
Skype:  luca.pal-robotics

Facebook http://www.facebook.com/palrobotics1 - Twitter
http://twitter.com/#%21/palrobotics - PAL Robotics YouTube Channel
http://www.youtube.com/user/PALRobotics

P Antes de imprimir este e-mail piense bien si es necesario hacerlo: El
medioambiente es cosa de todos.

AVISO DE CONFIDENCIALIDAD: Este mensaje y sus documentos adjuntos, pueden
contener información privilegiada y/o confidencial que está dirigida
exclusivamente a su destinatario.  Si usted recibe este mensaje y no es el
destinatario indicado, o el empleado encargado de su entrega a dicha
persona, por favor, notifíquelo inmediatamente y remita el mensaje original
a la dirección de correo electrónico indicada. Cualquier copia, uso o
distribución no autorizados de esta comunicación queda estrictamente
prohibida.

CONFIDENTIALITY NOTICE: This e-mail and the accompanying document(s) may
contain confidential information which is privileged and intended only for
the individual or entity to whom they are addressed.  If you are not the
intended recipient, you are hereby notified that any disclosure, copying,
distribution or use of this e-mail and/or accompanying document(s) is
strictly prohibited.  If you have received this e-mail in error, please
immediately notify the sender at the above e-mail address.

**Comment by bmagyar** [comment]
I also support this PR. Having a specific interface for setting current limits is preferred rather than emulating it with other (similar) interfaces where the mapping is not clear.

:+1:

**Comment by po1** [comment]
We should also update/revert these PRs:
- [ ] pal_control_msgs/CurrentLimit.srv (#9) I would just update this to use this new message
- [ ] pal_common_msgs/JointCurrent.msg (#10)  This one becomes redundant. It should be reverted and the code that uses it updated.

And lastly:
- [ ] rebase this PR

**Comment by adolfo-rt** [comment]
The PR has been rebased. I propose to act in this order, to ease the migration path:
- Merge this PR, as it currently stands.
- Commit some pending changes I have on our private repo to allow the coexistence of these changes with #9 and #10.
- Migrate code in trunk so it no longer uses the features in #9 and #10.
- Deprecate #9 and #10.
- Make new robot.

**Comment by po1** [comment]
> Merge this PR, as it currently stands.
> Commit some pending changes I have on our private repo to allow the coexistence of these changes with #9 and #10.
> Migrate code in trunk so it no longer uses the features in #9 and #10.
> Deprecate #9 and #10.
> Make new robot.
> - conquer world

**Pull request merged**